### PR TITLE
fix: Restore window after closing on MacOS

### PR DESCRIPTION
**Pull request by igor-sirotin**
Fixes https://github.com/status-im/status-desktop/issues/11405
Requires https://github.com/status-im/nimqml/pull/53

### 1. Use "method swizzling" technique instead of replacing ApplicationDelegate 
This is the actual fix. Left more details in comments:
https://github.com/status-im/status-desktop/blob/71f379dcf3f8dfacbe9a2849acf5aba856bcf7e1/vendor/DOtherSide/lib/src/Status/AppDelegate.mm#L52-L66

### 3. Move `DockShowEvent` logic to QML
Now code is much simpler.

https://github.com/status-im/status-desktop/blob/71f379dcf3f8dfacbe9a2849acf5aba856bcf7e1/ui/main.qml#L231-L240

### 3. Merged entitlements to a single file

Now all entitlements are signed as part of AppBundle.

## Screenshot of functionality (including design for comparison)

https://github.com/status-im/status-desktop/assets/25482501/ed574aa3-7050-49bf-8606-0b1373bdb45d

### CODE REVIEW

**Review by caybro — COMMENTED** [review]
Great stuff but there's still things to remove in `https://github.com/status-im/nimqml/blob/master/src/nimqml/private/status/statusevent.nim`

**Comment by igor-sirotin** [comment]
@caybro indeed, thank you!
I remember removing this code, but it seems that I dropped the changes in `nimqml` 😄 

https://github.com/status-im/nimqml/pull/53

**Review by caybro — APPROVED** [review]
Nice stuff :pinched_fingers: 

BTW have you tested on other platforms that the closing/quitting still works as expected? ;)

**Comment by igor-sirotin** [comment]
> BTW have you tested on other platforms that the closing/quitting still works as expected? ;)

@caybro of course not 😄 
Actually I don't think it affected any other OS, because all related code was under MacOS condition. Should be fine 🙂

File: ui/main.qml [hunk 1]
```diff
@@ -220,6 +225,22 @@ StatusWindow {
         }
     }
 
+    // On MacOS, explicitely restore the window on activating
+    Connections {
+        target: Qt.application
```
Comment (caybro) [inline ui/main.qml:230]: you could actually add a hint here:
```qml
enabled: Qt.platform.os === Constants.mac
```

Comment (igor-sirotin) [reply ui/main.qml:230]: image

File: ui/main.qml [hunk 2]
```diff
@@ -220,6 +225,22 @@ StatusWindow {
...
+        function onStateChanged() {
+            if (Qt.platform.os != Constants.mac)
+                return
```
Comment (igor-sirotin) [inline ui/main.qml:233]:
```suggestion
    Connections {
        target: Qt.application
        enabled: Qt.platform.os === Constants.mac
        function onStateChanged() {
```

File: ui/main.qml [hunk 3]
```diff
@@ -220,6 +225,21 @@ StatusWindow {
         }
     }
 
+    // On MacOS, explicitely restore the window on activating
+    Connections {
+        target: Qt.application
+        enabled: Qt.platform.os === Constants.mac
+        function onStateChanged() {
+
+            if (d.previousApplicationState == Qt.application.state) {
```
Comment (borismelnik) [inline ui/main.qml:234]: I can't understand it... so if `d.previousApplicationState == Inactive` and `Qt.application.state == Inactive` as well it should be shown normal? And be visible?

File: vendor/DOtherSide/lib/src/Status/AppDelegate.mm [hunk 4]
```diff
@@ -2,38 +2,70 @@
...
         QUrl deeplink = QUrl::fromNSURL(url);
         // set it to nim
         return TRUE;
     }
     return FALSE;
 }
 
+- (void)applicationDidFinishLaunching:(NSNotification *)notification
+{
+    qDebug() << "StatusApplicationDelegate::applicationDidFinishLaunching";
```
Comment (borismelnik) [inline vendor/DOtherSide/lib/src/Status/AppDelegate.mm:33]: is It should be here? maybe then `qInfo()`?

File: vendor/DOtherSide/lib/src/Status/AppDelegate.mm [hunk 5]
```diff
@@ -2,38 +2,70 @@
...
 namespace app_delegate {
 
+void swizzle_appdelegate_method(SEL selector) {
+    Class originalClass = [NSApplication sharedApplication].delegate.class;
+    Class swizzledClass = [StatusApplicationDelegate class];
+
+    Method originalMethod = class_getInstanceMethod(originalClass, selector);
+    Method swizzledMethod = class_getInstanceMethod(swizzledClass, selector);
+
+    method_exchangeImplementations(originalMethod, swizzledMethod);
```
Comment (borismelnik) [inline vendor/DOtherSide/lib/src/Status/AppDelegate.mm:47]: Awesome!

**Review by borismelnik — APPROVED** [review]
👍

File: ui/main.qml (hunk 3 above)
Comment (igor-sirotin) [reply ui/main.qml:234]: oh! good catch!
I was copying the logic from `DockShowEvent`
https://github.com/status-im/status-desktop/blob/bf4af3d7d4c32e0b9ecf71f6d8051b79996fcef0/vendor/DOtherSide/lib/src/Status/DockShowAppEvent.cpp#L27-L28

But lost half of it 😄 

I don't really understand why the state should be the same as previous, but I decided not to break it, as it was working fine.

**Pull request merged**